## Toilet problem :)
In our office there are a lot of people, and only one toilet with 4 cabins in there, but I've never waited in toilet! Everytime I come there is at least one cabin available.
So, I'd like to calculate:
1. Probability that I come to toilet and all cabins are busy
2. Number of days which I will wait for such an event
3. Number of people we should hire to meet such a problem every day

In [26]:
import scipy.stats as sts

In [27]:
# Healthy person goes to the toilet 5-7 times per day (let it be 6 in average), and one per night
# Let's think that night lasts 8 houra and day lasts 16 hours
# then during working day person attends toilet 3 times

# Number of toilet visits per workday
number_visit_toilet = 3.0

# Duration of workday in minutes
workday_duration_mins = 8 * 60.0

# Number of workers in company
number_workers = 100.0

# Number of cabins in toilet (pissuar == cabin)
number_cabins = 4.0

# Time interval that person spends in toilet (average)
duration_in_toilet_mins = 2.0 # minutes

# Calculate probability of event that person goes to toilet in definite period = duration_in_toilet_mins
prob_in_duration = number_visit_toilet * (duration_in_toilet_mins / workday_duration_mins)


In [28]:
# I think Poisson distribution is very good for such a calculations
# Calculate lambda for poisson distribution
poisson_lambda = prob_in_duration * number_workers

def probability_all_busy_in_toilet(poisson_lambda, number_cabins):
    # Init poisson distribution object
    poisson_rv = sts.poisson(poisson_lambda)
    return 1.0 - poisson_rv.cdf(number_cabins + 1.0)

print "Just fo fun, probability that nobody in toilet: ", sts.poisson(poisson_lambda).cdf(0)

busy_prob = probability_all_busy_in_toilet(poisson_lambda, number_cabins)
print "Probability that there are more persons in toilet than cabins available: ", busy_prob

Just fo fun, probability that nobody in toilet:  0.28650479686
Probability that there are more persons in toilet than cabins available:  0.00183808545059


In [24]:
print "Average number of days after which I will see all cabins busy: ", 1.0 / (busy_prob * number_visit_toilet)

Average number of days after which I will see all cabins busy:  181.348115903


In [25]:
# Let's calculate how many people should work in the company to cause this kind of problem on daily basis
for number_workers in range(100, 1000, 1):
    busy_prob = probability_all_busy_in_toilet(prob_in_duration * number_workers, number_cabins)
    if busy_prob * number_visit_toilet > 1.0:
        print "How many people should work: ", number_workers
        break

How many people should work:  377


## Conclusions:
1. Every third time I attend toilet - all cabins are empty. Cool!
2. Probability to see all cabins busy - 0.00183808545059. Rather small, but not that understandable.
3. Probably I will see all cabins busy twice per year - not that bad.
4. Situation will get worse when we hire more, but we should hire another 277 workers to meet that unpleasent situation every day.
So, I can relax :)